server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -303,17 +310,130 @@ private CoordinationDiagnosticsResult diagnoseOnHaveNotSeenMasterRecently(Master
         } else if (leaderHasBeenElected) {
             DiscoveryNode currentMaster = coordinator.getPeerFinder().getLeader().get();
             result = getResultOnCannotJoinLeader(localMasterHistory, currentMaster, explain);
-        } else {
+        } else if (clusterService.localNode().isMasterNode() == false) { // none is elected master and we aren't master eligible
             // NOTE: The logic in this block will be implemented in a future PR
             result = new CoordinationDiagnosticsResult(
                 CoordinationDiagnosticsStatus.RED,
                 "No master has been observed recently",
                 CoordinationDiagnosticsDetails.EMPTY
             );
+        } else { // none is elected master and we are master eligible
+            result = diagnoseOnHaveNotSeenMasterRecentlyAndWeAreMasterEligible(localMasterHistory, masterEligibleNodes, explain);
+        }
```

Comment:
I think this if/else block is becoming hard to follow and reason about 
ie. where do we check _we_ aren't master eligibile node? the last `else` statement has a bunch of implicit decisions that are hard to verify (ie. why are we sure we're master eligible in this case?)

Reply:
The `else if` right above the else checks if we're master eligible. I'll try to make it more explicit.